modules/matrix_controller.py

```
@@ -662,139 +664,395 @@ def apply_pattern(self, pattern, color, brightness, speed):
+                self.is_streaming = True
+                self.animation_thread = threading.Thread(
+                    target=self._matrix_rain_animation_loop,
+                    args=(speed,),
+                    daemon=True
+                )
+                self.animation_thread.start()
+                return True
+
+            elif pattern == "custom":
+                # Handle custom pattern data
+                custom_data = data.get("data", [])
```

Ruff (0.12.2)

Undefined name `data`

(F821)